### DISCUSSION TITLE
Stereochemistry in Ring-opening reactions

### DISCUSSION THREAD

**Discussion by levineds** [rdkit/rdkit#4401] [Q&A]
I am running a ring-opening reaction with reaction SMARTS, and I'm having a problem with the stereochemistry of the product. The reactant can be a small ring that could possibly contain a double bond. Since it's a small ring, the bond is obviously cis, RDKit doesn't even allow one to specify that it is cis since it's "obvious". However, once the product is formed, there is no longer a ring and the double bond now has undefined stereochemistry and arbitrarily comes out trans (which is wrong, it started out cis). 

Here is a minimal sort of example
mol = Chem.MolFromSmiles('O1NC=CC1')
rxn = AllChem.ReactionFromSmarts('[O:2][N:3]>>([O:2].[N:3])')
products = rxn.RunReactants([mol])
IPythonConsole.ShowMols(products[0])

Reaction SMARTS are supposed to honor stereochemistry in unspecified atoms, but because small rings don't support stereochemistry of double bonds, there isn't even any stereochemistry for the product to check. How do I tell RDKit that it needs to preserve the cis-ness of the reactant double bond in the product?

**Comment by d-b-w** [comment]
Here's what that looks like:
<img width="800" alt="ring_opening_reaction" src="https://user-images.githubusercontent.com/3013277/128577270-5438d1cf-a39f-4765-9fa7-d195128705fb.png">

**Comment by greglandrum** [comment]
You've diagnosed what is going on perfectly: the RDKit doesn't assign bond stereo labels to bonds which are in small rings.

It is, however, possible to set the stereo on the double bond after the molecule is constructed, and this information will then be preserved:
```
>>> mol = Chem.MolFromSmiles('O1NC=CC1')
>>> rxn = AllChem.ReactionFromSmarts('[O:2][N:3]>>([O:2].[N:3])')
>>> products = rxn.RunReactants([mol])
>>> Chem.SanitizeMol(products[0][0])
rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE
>>> products[0][0].Debug()
Atoms:
	0 8 O chg: 0  deg: 1 exp: 1 imp: 1 hyb: 4 arom?: 0 chi: 0
	1 7 N chg: 0  deg: 1 exp: 1 imp: 2 hyb: 3 arom?: 0 chi: 0
	2 6 C chg: 0  deg: 2 exp: 2 imp: 2 hyb: 4 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	4 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
Bonds:
	0 2->0 order: 1 conj?: 0 aromatic?: 0
	1 3->2 order: 1 conj?: 0 aromatic?: 0
	2 4->3 order: 2 conj?: 1 aromatic?: 0
	3 1->4 order: 1 conj?: 1 aromatic?: 0
>>> bnd = mol.GetBondBetweenAtoms(2,3)
>>> bnd.SetStereoAtoms(1,4)
>>> bnd.SetStereo(Chem.BondStereo.STEREOCIS)
>>> products = rxn.RunReactants([mol])
>>> Chem.SanitizeMol(products[0][0])
rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE
>>> products[0][0].Debug()
Atoms:
	0 8 O chg: 0  deg: 1 exp: 1 imp: 1 hyb: 4 arom?: 0 chi: 0
	1 7 N chg: 0  deg: 1 exp: 1 imp: 2 hyb: 3 arom?: 0 chi: 0
	2 6 C chg: 0  deg: 2 exp: 2 imp: 2 hyb: 4 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	4 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
Bonds:
	0 2->0 order: 1 conj?: 0 aromatic?: 0
	1 3->2 order: 1 conj?: 0 aromatic?: 0
	2 4->3 order: 2 stereo: 4 stereoAts: (1 2) conj?: 1 aromatic?: 0
	3 1->4 order: 1 conj?: 1 aromatic?: 0
>>> Chem.SetDoubleBondNeighborDirections(products[0][0])
>>> Chem.MolToSmiles(products[0][0])
'N/C=C\\CO'
```

Obviously that's a one-off, but if you regularly faced this problem you could write a function which finds all double bonds in small rings and sets the stereo of those bonds to STEREOCIS.

**Reply by levineds** [reply]
Thanks for clarifying that I have diagnosed the problem correctly. This seems then to be a bug in the RunReactants method; according to the RDKit documentation "If no chiral information is present in the reaction definition, the stereochemistry of the reactants is preserved...In general, the reaction machinery tries to preserve as much stereochemistry information as possible." The cis nature of the double-bond in the small ring is unambiguous (indeed, that is why sanitizing a molecule with such a label removes the label). Therefore, it should be unambiguously transferred automatically to the product. There is no case in which the current behavior could be correct in the paradigm described in the documentation (i.e. that if no stereochemistry is specified, preserve whatever currently exists).

**Comment by bp-kelley** [comment]
Sorry, message accidentally sent, trying again:

I think you misread what Greg wrote.  The RDKit never assigns double bond stereo to small rings, so RunReactants simply can’t copy over the non-existent data.

The example he gave:

>>> bnd = mol.GetBondBetweenAtoms(2,3)
>>> bnd.SetStereoAtoms(1,4)
>>> bnd.SetStereo(Chem.BondStereo.STEREOCIS)
Adds the stereo information to the bond that RunReactants will copy to the product.

You can loop through the reactant’s bonds and do something like:

for bond in mol.GetBonds():
  if bond.GetBondTypeAsDouble() == 2 and bond.IsInRing():
      bond.SetStereo(Chem.BondStereo.STEREOCIS)
     

Hopefully this solves your issue.
----
Brian Kelley

**Reply by levineds** [reply]
No, I believe I understood it correctly.

We agree that RDKit never assigns double bond stereo in small rings, the reason being that such labels are extraneous; the bonds must be cis and we all (including RDKit) know that. As a result, if the paradigm that all known stereochemistry is transferred is to be honored, then, internal to RunReactants, the cis nature of the in-ring double bond must be transferred to the product. This is the bug: the method has forgotten about the implicit nature of cis double bonds in small rings, naively checking only for a stereo label. The label is missing from the reactants not because the stereochemistry is ambiguous or unknown, but precisely because it is not; the bond must be cis. To be correct, the product must also be cis (unless specified otherwise as with other stereochemistry in reaction SMARTS); hence the code is buggy.

I appreciate the temporary workaround, but the current state is definitively incorrect.

For completeness also, the workaround loop should be (with appropriate indents):
for bond in mol.GetBonds():
   if bond.GetBondTypeAsDouble() == 2 and any(bond.IsInRingSize(i) for i in range(3,8)):
      bond.SetStereo(Chem.BondStereo.STEREOCIS)

**Reply by greglandrum** [reply]
@levineds : I don't necessarily agree that this is a bug, since the code is doing what it was designed to do, but I certainly think it would be nice if the behavior here were different.
I'll try and make some time to take a look and investigate changing the behavior

**Reply by levineds** [reply]
Thanks @greglandrum. Also I spoke to soon about the code I posted above. You need to also find the atoms that come before and after the double bond in the ring so that you can set stereo atoms.
Here is a working snippet of code:
```
ring_info = mol.GetRingInfo()                                  
for ring in ring_info.BondRings():                             
    if len(ring) > 7:                                          
        continue                                               
    for ring_bond_idx, bond_idx in enumerate(ring):            
        bond = mol.GetBondWithIdx(bond_idx)                    
        if bond.GetBondTypeAsDouble() != 2:                    
            continue                                           
        next_bond = mol.GetBondWithIdx(ring[(ring_bond_idx+1)%len(ring)])  
        prev_bond = mol.GetBondWithIdx(ring[ring_bond_idx-1])
        bond_ats = {bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()}
        next_at = ({next_bond.GetBeginAtomIdx(), next_bond.GetEndAtomIdx()} - bond_ats).pop()
        prev_at = ({prev_bond.GetBeginAtomIdx(), prev_bond.GetEndAtomIdx()} - bond_ats).pop()
        if bond.GetBeginAtomIdx() in {prev_bond.GetBeginAtomIdx(), prev_bond.GetEndAtomIdx()}:
            bond.SetStereoAtoms(prev_at, next_at)
        else:
            bond.SetStereoAtoms(next_at, prev_at)
        bond.SetStereo(Chem.BondStereo.STEREOCIS)
